### PR TITLE
TEST-19842 Automated Parallel Copy Operation of same source and destination

### PR DESCRIPTION
**Pull request by ketanarlulkar**
# Problem Statement
- Automated test defined at _Table2 : Parallel Copy Operation of same source and destination_ in [TDS: F-20A: S3 Copy Object API](https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/1079640340/TDS+F-20A+S3+Copy+Object+API)

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [x] New/Affected tests are executed on Latest Build
-  [x] Attach test execution logs ([Logs](https://jts.seagate.com/browse/CORTX-33459?focusedCommentId=2405947&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-2405947))
-  [x] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: tests/s3/copy_object/test_parallel_copy_object.py [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+
+    @pytest.mark.tags("TEST-44792")
+    @CTFailOn(error_handler)
+    def test_44792(self):
+        """Test Parallel Self copy and put object in same bucket"""
+        LOGGER.info("STARTED: Test Parallel Self copy and put object in same bucket "
+                    "(simple and multipart source object)")
+
+        LOGGER.info("Upload object1 to source bucket1")
+        self.create_put_object("src-bucket", "obj1", 100)
```
Comment (archanabpatil) [inline tests/s3/copy_object/test_parallel_copy_object.py:260]: make this 5gb object for variation

File: tests/s3/copy_object/test_parallel_copy_object.py [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+
+        self.parallel_copy_and_put_object(
+            ("src-bucket", "obj1", "src-bucket", "obj1", "InvalidRequest"),
+            ("src-bucket", "obj2", self.file_path))
+
+        LOGGER.info("Upload multipart object obj3 to src-bucket")
+        self.s3mp_obj.simple_multipart_upload("src-bucket", "obj3", 1024, self.file_path, 4)
```
Comment (archanabpatil) [inline tests/s3/copy_object/test_parallel_copy_object.py:267]: lets make it 3 gb object and rename it to mp-obj1 like above tests. I will modify name in tds and test

**Review by archanabpatil — CHANGES_REQUESTED** [review]

**Comment by archanabpatil** [comment]
I have updated the object names wherever I had asked in this PR to tds and in tests

File: tests/s3/copy_object/test_parallel_copy_object.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+        """Test Parallel copy and put operation from same source bucket"""
+        LOGGER.info("STARTED: Test Parallel copy and put operation from same source bucket "
+                    "(for simple and multipart source objects)")
+        LOGGER.info("Upload object1 to source bucket1")
+        self.create_put_object("src-bucket", "obj1", 100)
+        self.parallel_copy_and_put_object(("src-bucket", "obj1", "dest-bucket", "obj2"),
+                                          ("src-bucket", "obj3", self.file_path))
+        LOGGER.info("Upload multipart object mp-obj1 to src-bucket")
+        self.s3mp_obj.simple_multipart_upload("src-bucket", "mp-obj1", 1024, self.file_path, 4)
+
```
Comment (archanabpatil) [inline tests/s3/copy_object/test_parallel_copy_object.py:131]: blank line

**Review by archanabpatil — APPROVED** [review]

File: libs/s3/s3_multipart_test_lib.py [hunk 4]
```diff
@@ -441,7 +441,8 @@ def get_byte_range_of_object(self, bucket_name: str = None, my_key: str = None,
     def simple_multipart_upload(self, bucket_name: str, object_name: str, file_size: int,
                                 file_path: str, parts: int):
         """
-        Do multipart upload for given object by dividing it into given parts.
+        Create the local file of size file_size*MiB at file_path. Do multipart upload of file_path
+        by dividing it into given number of parts.
```
Comment (AshwiniS-Borse) [inline libs/s3/s3_multipart_test_lib.py:445]: Need to add return param as well.

**Review by AshwiniS-Borse — APPROVED** [review]

File: tests/s3/copy_object/test_parallel_copy_object.py [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+        LOGGER.info("Parallely, \n1. From %s copy %s to %s as %s\n2. Put %s to %s as %s",
+                    copy_args[0], copy_args[1], copy_args[2], copy_args[3],
+                    put_args[2], put_args[0], put_args[1])
+        with multiprocessing.Pool(processes=2) as pool:
+            process1 = pool.apply_async(self.copy_object_wrapper, args=copy_args)
+            process2 = pool.apply_async(self.put_object_wrapper, args=put_args)
+            assert_utils.assert_true(process1.get()[0], process1.get()[1])
+            assert_utils.assert_true(process2.get()[0], process2.get()[1])
+
+    @pytest.mark.tags("TEST-44786")
```
Comment (ravindrachoudhari) [inline tests/s3/copy_object/test_parallel_copy_object.py:119]: no other marker required?

Comment (ketanarlulkar) [reply tests/s3/copy_object/test_parallel_copy_object.py:121]: Added markers

**Review by ravindrachoudhari — APPROVED** [review]

**Pull request merged**